Encryption.md

```diff
@@ -19,216 +19,544 @@
+   * Dictionary Page (3)
+   * Data PageHeader (4)
+   * Dictionary PageHeader (5)
+   * ColumnIndex (6)
+   * OffsetIndex (7)
+
+
+|                      |File ID     | Module type | Row group ordinal | Column ordinal | Page ordinal|
+|----------------------|------------|-------------|-------------------|----------------|-------------|
+| Footer               |     V      |    V (0)    |         X         |       X        |      X      |
+| ColumnMetaData       |     V      |    V (1)    |         V         |       V        |      X      |
+| Data Page            |     V      |    V (2)    |         V         |       V        |      V      |
+| Dictionary Page      |     V      |    V (3)    |         V         |       V        |      X      |
+| Data PageHeader      |     V      |    V (4)    |         V         |       V        |      V      |
+| Dictionary PageHeader|     V      |    V (5)    |         V         |       V        |      X      |
+| ColumnIndex          |     V      |    V (6)    |         V         |       V        |      X      |
+| OffsetIndex          |     V      |    V (7)    |         V         |       V        |      X      |
+
+
+
+## 5. File Format
+Parquet file encryption algorithm is specified in a union of the following Thrift structures:
+
+```c
+struct AesGcmV1 {
+  /** Retrieval metadata of AAD prefix **/
+  1: optional binary aad_metadata
+
+  /** Unique file identifier part of AAD suffix **/
+  2: optional binary aad_file_unique
+}
+
+struct AesGcmCtrV1 {
+  /** Retrieval metadata of AAD prefix **/
+  1: optional binary aad_metadata
+
+  /** Unique file identifier part of AAD suffix **/
+  2: optional binary aad_file_unique
+}
+
+union EncryptionAlgorithm {
+  1: AesGcmV1 AES_GCM_V1
+  2: AesGcmCtrV1 AES_GCM_CTR_V1
+}
+```
+ 
+
+The Thrift modules are encrypted after serialization, using the GCM cipher. For each module, the encryption 
+buffer is comprised of a nonce, ciphertext and tag, described in the Algorithms section. The column pages 
+(data and dictionary) are encrypted after compression. With the AES_GCM_V1 algorithm, a page encryption 
+buffer is comprised of a nonce, ciphertext and tag. The length of the encryption buffer 
+(a 4-byte little endian) is written to the output stream, followed by the buffer itself.
+
+|length (4 bytes) | nonce (12 bytes) | ciphertext (length-28 bytes) | tag (16 bytes) |
+|-----------------|------------------|------------------------------|----------------|
+
+
+With the AES_GCM_CTR_V1 algorithm, a page encryption buffer is comprised of a nonce and ciphertext, 
+described in the Algorithms section. The length of the encryption buffer 
+(a 4-byte little endian) is written to the output stream, followed by the buffer itself.
+
+|length (4 bytes) | nonce (12 bytes) | ciphertext (length-12 bytes) |
+|-----------------|------------------|------------------------------|
+
+
+A `crypto_meta_data` field is set in each ColumnChunk in the encrypted columns. ColumnCryptoMetaData 
```

Comment:
I thought that the decision was to use "metadata" instead of "meta_data"?

Reply:
parquet-cpp has released the version 1.5.0 with these crypto structures.
https://github.com/apache/parquet-cpp/blob/apache-parquet-cpp-1.5.0/src/parquet/parquet.thrift#L708

While this version cant read files with encrypted footer, or encrypted columns in files with plaintext footer - it should be able to read plaintext columns in encrypted files with plaintext footer - as any legacy reader.

However, this means we cant rename a couple of fields, including this one. The cpp thrift above has:
```c
struct ColumnChunk {
  8: optional ColumnCryptoMetaData crypto_meta_data
```

Reply:
I thought the point of field IDs was that we can rename fields. It may just work.

As for the 1.5.0 release of parquet-cpp, I don't think we intend to support files written with the encryption structures from that release. The spec was not approved.

And aren't we making a lot of other changes to the spec anyway? I would expect more than just this to break. Column metadata, for example, has moved.

Reply:
I think IDs should work here.